tensorboard/uploader/uploader.py

```diff
@@ -224,52 +235,81 @@ def send_requests(self, run_to_events):
+        # ignored with a warning.
+        metadata = self._tag_metadata.get(time_series_key)
+        if metadata is None:
+            metadata = value.metadata
+            self._tag_metadata[time_series_key] = metadata
+
+        # TODO(soergel): Allow enabling/disabling upload per plugin
+
+        if value.HasField("metadata") and (
+            value.metadata.plugin_data.plugin_name
+            != metadata.plugin_data.plugin_name
+        ):
+            logger.warning(
+                "Mismatching plugin names for %s.  Expected %s, found %s.",
+                time_series_key,
+                metadata.plugin_data.plugin_name,
+                value.metadata.plugin_data.plugin_name,
+            )
+        elif metadata.data_class == summary_pb2.DATA_CLASS_SCALAR:
+            self._scalar_request_sender.add_event(
+                run_name, event, value, metadata
+            )
+        # TODO(nielsene): add Tensor sender
+        # elif metadata.data_class == summary_pb2.DATA_CLASS_TENSOR:
+        #     self._tensor_request_sender.add_event(
+        #         run_name, event, value, metadata
+        #     )
+        elif metadata.data_class == summary_pb2.DATA_CLASS_BLOB_SEQUENCE:
+            self._blob_request_sender.add_event(
+                run_name, event, value, metadata
+            )
 
-    def _run_values(self, run_to_events):
-        """Helper generator to create a single stream of work items."""
-        # Note that each of these joins in principle has deletion anomalies:
-        # if the input stream contains runs with no events, or events with
-        # no values, we'll lose that information. This is not a problem: we
-        # would need to prune such data from the request anyway.
+    def _run_events(self, run_to_events):
+        """Helper generator to create a single stream of work items.
+
+        The events are passed through the `data_compat` and `dataclass_compat`
+        layers before being emitted, so downstream consumers may process them
+        uniformly.  Note that `dataclass_compat` may emit multiple variants of
+        the same event, for backwards compatibility.  Thus this stream should
+        be filtered to obtain the desired version of each event.  Here, this
+        happens below in `_summary_values(...)`, which ignores any event that
+        does not have a `summary` field.
+        """
+        # Note that this join in principle has deletion anomalies: if the input
+        # stream contains runs with no events, we'll lose that information. This
+        # is not a problem: we would need to prune such data from the request
+        # anyway.
         for (run_name, events) in six.iteritems(run_to_events):
             for event in events:
-                for value in event.summary.value:
-                    yield (run_name, event, value)
+                v2_event = data_compat.migrate_event(event)
+                dataclass_events = dataclass_compat.migrate_event(v2_event)
```

Comment:
Note that this *is* a behavior change for existing flows. Once this is
released, metadata for scalar time series will have the `data_class`
field set where previously it did not. This is fine, but tightens that
data compatibility constraint, so we should release it intentionally.

Reply:
I agree that here we set data_class where previously we didn't.  I'm not sure I follow your concern, but I think it is that server-side we have to make sure things work whether the field is set or not.  Indeed, in the associated internal changes, we do fall back to scalars when data_class is empty, but that is worth doublechecking in tests and reviews.

Reply:
The concern/observation was that this is the first place that a proto
with `data_class` set is actually written to any persistent store, which
reduces our flexibility to change the structure or interpretation (but
not to zero, because we’re still only writing it to TensorBoard.dev, not
users’ local disks). No action required.